src/org/omegat/core/search/Searcher.java

```diff
@@ -628,7 +630,24 @@ public boolean searchString(String origText, boolean collapseResults) {
             }
             while (true) {
                 int start = matcher.start();
-                foundMatches.add(new SearchMatch(start, matcher.end()));
+                if (m_searchExpression.mode == SearchMode.REPLACE && m_searchExpression.searchExpressionType == SearchExpression.SearchExpressionType.REGEXP) {
+                    String repl = m_searchExpression.replacement;
+                    Matcher replaceMatcher = Pattern.compile("(?<!\\\\)\\$(\\d+)").matcher(repl);
+                    while (replaceMatcher.find()) {
+                        int varId = Integer.parseInt(replaceMatcher.group(1));
+                        repl = repl.substring(0, replaceMatcher.start())
+                            + matcher.group(varId) // yes, from source matcher!
```

Comment:
I'm sure you did not consider this to be within the scope of your patch, but for completeness: when the user specifies a group that doesn't exist (e.g. search: `.` replace: `$1`) the error message has some problems:

![screen shot 2018-02-04 at 10 58 33](https://user-images.githubusercontent.com/2172537/35773595-638acae6-099a-11e8-9907-5ba4aecd6c0f.png)

1. The message is confusing: Technically, yes, we are searching a directory tree, but this is a detail the user doesn't really know about.
2. The part about groups may be understandable to a power user using regex replacements, but it could be improved
3. It appears parented to the main window, when it would be better on the search window
4. It appears even if we merely searched and did not initiate a replacement

At the very least we should catch the OOB exception and provide a better error message.

Reply:
Yes, I simply forgot to check that case. No problem to correct this.

Reply:
OK I could move the error message inside SearchThread and find better text (points 1 & 2)
But now it is modal, and difficult to say whenever it refers to search window or main window